### DISCUSSION TITLE
How to send INSERT/UPDATE without CTE for PostgreSQL?

### DISCUSSION THREAD

**Discussion by kevinelliott** [dart-backend/angel#150] [Q&A]
Currently when inserting or updating a record, the ORM is issuing the query as a CTE (WITH ... UPDATE ... SELECT). This impedes the ability to send reads to replicas via the front pooler (pgcat) because it breaks the transaction, resulting with errors like:
```
Ingest(acarsdec) #1110: [ERROR] - [acars] Unable to update flight: Severity.error 25006: cannot execute SELECT in a read-only transaction
```

Solution for this would be to allow INSERT/UPDATE without CTE style, with a methodology something like:
1. Take original ORM originated object and update the values in the object, or construct a new object
2. Pass the object to insert/update functions
3. If success, no need to SELECT the content and return an object -- as we have the data already in the object.
4. If fail, raise the exception (and thus the object is invalid state).

This would allow for UPDATEs and INSERTs to be sent without CTE.

**Comment by dukefirehawk** [comment]
Currently CTE is used everywhere within ORM. It is used extensively to enable multiple queries to be optimally executed by the database engine. The SELECT is mainly used to return the ID generated by auto increment primary key. Dropping the SELECT would break this logic. I'll take a look if there is a better way to address this besides dropping the CTE which would require a major rework.

**Reply by kevinelliott** [reply]
Thanks @dukefirehawk, looking forward to it.

I have also submitted an issue on the pgcat repo (https://github.com/postgresml/pgcat/issues/829) to request CTE detection in order to not send the select to a replica, perhaps as a stop gap measure (assuming they are willing to do this). Though hopefully we can figure out something here!
